blockchain/chain_sync/src/network_handler.rs

```diff
@@ -3,47 +3,51 @@
 
 use super::peer_manager::PeerManager;
 use async_std::prelude::*;
-use async_std::sync::{Receiver, Sender};
+use async_std::sync::Mutex;
+use async_std::sync::Receiver;
 use async_std::task;
+use flo_stream::{MessagePublisher, Publisher};
 use forest_libp2p::rpc::{RPCResponse, RequestId};
 use forest_libp2p::NetworkEvent;
+use futures::channel::oneshot::Sender as OneShotSender;
 use log::trace;
+use std::collections::HashMap;
 use std::sync::Arc;
 
-pub(crate) type RPCReceiver = Receiver<(RequestId, RPCResponse)>;
-pub(crate) type RPCSender = Sender<(RequestId, RPCResponse)>;
-
 /// Handles network events from channel and splits based on request
 pub(crate) struct NetworkHandler {
-    rpc_send: RPCSender,
-    event_send: Sender<NetworkEvent>,
+    event_send: Publisher<NetworkEvent>,
     receiver: Receiver<NetworkEvent>,
+    /// keeps track of a mapping from rpc request id to oneshot senders
+    request_table: Arc<Mutex<HashMap<RequestId, OneShotSender<RPCResponse>>>>,
 }
 
 impl NetworkHandler {
     pub(crate) fn new(
         receiver: Receiver<NetworkEvent>,
-        rpc_send: RPCSender,
-        event_send: Sender<NetworkEvent>,
+        event_send: Publisher<NetworkEvent>,
+        request_table: Arc<Mutex<HashMap<RequestId, OneShotSender<RPCResponse>>>>,
     ) -> Self {
         Self {
             receiver,
-            rpc_send,
             event_send,
+            request_table,
         }
     }
 
     pub(crate) fn spawn(&self, peer_manager: Arc<PeerManager>) {
         let mut receiver = self.receiver.clone();
-        let rpc_send = self.rpc_send.clone();
-        let event_send = self.event_send.clone();
+        let mut event_send = self.event_send.republish();
+        let request_table = self.request_table.clone();
 
         task::spawn(async move {
             loop {
                 match receiver.next().await {
                     // Handle specifically RPC responses and send to that channel
                     Some(NetworkEvent::RPCResponse { req_id, response }) => {
-                        rpc_send.send((req_id, response)).await
+                        // look up the request_table for the id and send through channel
+                        let tx = request_table.lock().await.remove(&req_id).unwrap();
+                        tx.send(response).unwrap();
```

Comment:
kinda same here, not exactly familiar when the oneshot channel can fail, but I'm thinking in this case it is probably better for this to be an async-std channel because the responses are potentially very large and would be better for it to be async maybe?

Reply:
This fails if the receiver is dropped, which it shouldn't be, but ill change it to log an error is something happens here.